File: backend/app/schemas/paper.py
# Pydantic schemas for paper list, detail, upload, processing, and overview payloads.

from datetime import datetime
from pydantic import BaseModel, ConfigDict
from typing import List, Optional

from app.schemas.paragraph import ParagraphResult, ElementResponse


# Full ORM-backed paper response payload.
class PaperResponse(BaseModel):
    id: int
    title: str | None
    original_filename: str
    stored_file_path: str
    parse_status: str
    parse_error: str | None = None
    parse_started_at: str | None = None
    parse_finished_at: str | None = None
    overview_status: str
    overview_error: str | None = None
    overview_started_at: str | None = None
    overview_finished_at: str | None = None
    zh_translation_status: str
    zh_translation_error: str | None = None
    export_status: str
    export_error: str | None = None
    export_started_at: str | None = None
    export_finished_at: str | None = None
    last_error_message: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# Compact paper row used by the home page list.
class PaperListItemResponse(BaseModel):
    paper_id: int
    title: str | None
    original_filename: str
    parse_status: str
    parse_error: str | None = None
    parse_started_at: str | None = None
    parse_finished_at: str | None = None
    overview_status: str
    overview_error: str | None = None
    overview_started_at: str | None = None
    overview_finished_at: str | None = None
    zh_translation_status: str
    zh_translation_error: str | None = None
    zh_translation_started_at: str | None = None
    zh_translation_finished_at: str | None = None
    export_status: str
    export_error: str | None = None
    export_started_at: str | None = None
    export_finished_at: str | None = None
    last_error_message: str | None = None


# Reader-page paper detail payload with structured elements.
class PaperDetailResponse(BaseModel):
    paper_id: int
    title: str | None
    original_filename: str
    parse_status: str
    parse_error: str | None = None
    parse_started_at: str | None = None
    parse_finished_at: str | None = None
    overview_status: str
    overview_error: str | None = None
    overview_started_at: str | None = None
    overview_finished_at: str | None = None
    zh_translation_status: str
    zh_translation_error: str | None = None
    zh_translation_started_at: str | None = None
    zh_translation_finished_at: str | None = None
    export_status: str
    export_error: str | None = None
    export_started_at: str | None = None
    export_finished_at: str | None = None
    last_error_message: str | None = None
    pdf_url: str
    elements: List[ElementResponse]


# Legacy processing response shape kept for compatibility.
class PaperProcessResponse(BaseModel):
    paper_id: int
    original_filename: str
    stored_file_path: str
    parse_status: str
    parse_error: str | None = None
    parse_started_at: str | None = None
    parse_finished_at: str | None = None
    overview_status: str
    overview_error: str | None = None
    paragraphs: list[ParagraphResult]


# Upload response shape returned after queueing parse work.
class UploadResponse(BaseModel):
    paper_id: int
    title: str | None
    original_filename: str
    parse_status: str
    parse_error: str | None = None
    parse_started_at: str | None = None
    parse_finished_at: str | None = None
    overview_status: str
    overview_error: str | None = None
    overview_started_at: str | None = None
    overview_finished_at: str | None = None
    zh_translation_status: str | None = None
    zh_translation_error: str | None = None
    export_status: str | None = None
    export_error: str | None = None
    export_started_at: str | None = None
    export_finished_at: str | None = None
    last_error_message: str | None = None
    pdf_url: str
    elements: List[ElementResponse]


# Paper schema highlight summary item.
class HighlightSummaryResponse(BaseModel):
    element_id: int
    title: str
    summary: str


# Paper schema section summary item.
class SectionSummaryResponse(BaseModel):
    section_key: Optional[str] = None
    section_title: str
    summary: str


# Paper overview response returned by overview endpoints.
class PaperOverviewResponse(BaseModel):
    paper_id: int
    language: str
    abstract_summary: str
    overall_summary: str
    overall_key_points: List[str]
    highlight_element_ids: List[int]
    highlight_summaries: List[HighlightSummaryResponse]
    section_summaries: List[SectionSummaryResponse]

File: frontend/src/types/paper.ts
// Frontend paper, element, PDF location, and overview types.

// Home-page paper list item returned by the backend.
export type PaperListItem = {
  paper_id: number;
  title: string | null;
  original_filename: string;
  parse_status: string;
  parse_error?: string | null;
  parse_started_at?: string | null;
  parse_finished_at?: string | null;
  overview_status: string;
  overview_error?: string | null;
  overview_started_at?: string | null;
  overview_finished_at?: string | null;
  zh_translation_status: string;
  zh_translation_error?: string | null;
  zh_translation_started_at?: string | null;
  zh_translation_finished_at?: string | null;
  export_status?: string | null;
  export_error?: string | null;
  export_started_at?: string | null;
  export_finished_at?: string | null;
  last_error_message?: string | null;
};

// Normalized PDF location used to jump/highlight generated paragraph matches.
export type PdfLocation = {
  page: number;
  bbox: [number, number, number, number];
};

// Reader-page paper detail payload.
export type PaperDetail = {
  paper_id: number;
  title: string | null;
  original_filename: string;
  parse_status: string;
  parse_error?: string | null;
  parse_started_at?: string | null;
  parse_finished_at?: string | null;
  overview_status: string;
  overview_error?: string | null;
  overview_started_at?: string | null;
  overview_finished_at?: string | null;
  zh_translation_status: string;
  zh_translation_error?: string | null;
  zh_translation_started_at?: string | null;
  zh_translation_finished_at?: string | null;
  export_status?: string | null;
  export_error?: string | null;
  export_started_at?: string | null;
  export_finished_at?: string | null;
  last_error_message?: string | null;
  pdf_url: string;
  elements: Element[];
};

// Structured reader element displayed in the paragraph table.
export type Element = {
  id: number;
  paragraph_id: number;
  type: "heading" | "paragraph" | "bullet_list";

  text?: string | null;
  summary?: string | null;
  key_points?: string[] | null;

  level?: string | null;

  intro_text?: string | null;
  items?: string[] | null;
  section_key?: string | null;

  // 新增：PDF 定位資訊
  page_number?: number | null;
  pdf_rects?: number[][];
  pdf_locations?: PdfLocation[];
};

// Paper-level highlight summary item.
export type HighlightSummary = {
  element_id: number;
  title: string;
  summary: string;
};

// Paper-level section summary item.
export type SectionSummary = {
  section_key?: string;
  section_title: string;
  summary: string;
};

// Paper-level overview payload.
export type PaperOverview = {
  paper_id: number;
  language: string;
  abstract_summary: string;
  overall_summary: string;
  overall_key_points: string[];
  highlight_element_ids: number[];
  highlight_summaries: HighlightSummary[];
  section_summaries: SectionSummary[];
};
